pkg/snapshot/restorer/restorer.go

```diff
@@ -887,3 +944,86 @@ func ErrorArrayToError(errs []error) error {
 
 	return fmt.Errorf("%s", strings.TrimSpace(errString))
 }
+
+// HandleAlarm function handles alarm raised by backup-restore.
+func (r *Restorer) HandleAlarm(stopHandleAlarmCh chan bool, dbSizeAlarmCh <-chan string, dbSizeDisAlarmCh chan bool, clientMaintenance client.MaintenanceCloser) {
+	r.logger.Info("Starting to handle an alarm...")
+	for {
+		select {
+		case <-stopHandleAlarmCh:
+			r.logger.Info("Closing handleAlarm...")
+			return
+		case endPoint := <-dbSizeAlarmCh:
+			r.logger.Info("Received a dbsize alarm")
+			r.logger.Infof("Calling defrag on endpoint: [%v]", endPoint)
+			if err := func() error {
+				ctx, cancel := context.WithTimeout(context.Background(), etcdDefragTimeout)
+				defer cancel()
+				if _, err := clientMaintenance.Defragment(ctx, endPoint); err != nil {
+					return err
+				}
+				return nil
+			}(); err != nil {
+				r.logger.Errorf("unable to disalarm as defrag call failed: %v", err)
+				// failed to disalarm
+				dbSizeDisAlarmCh <- false
+			} else {
+				// successfully disalarm
+				dbSizeDisAlarmCh <- true
+			}
+		}
+	}
+}
+
+// MakeEtcdLeanAndCheckAlarm calls etcd compaction on given revision number and raise db size alarm if embedded etcd db size crosses threshold.
+func (r *Restorer) MakeEtcdLeanAndCheckAlarm(revision int64, endPoints []string, embeddedEtcdQuotaBytes float64, dbSizeAlarmCh chan string, dbSizeDisAlarmCh <-chan bool, clientKV client.KVCloser, clientMaintenance client.MaintenanceCloser) error {
+	if err := func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), etcdCompactTimeout)
+		defer cancel()
+		if _, err := clientKV.Compact(ctx, revision, clientv3.WithCompactPhysical()); err != nil {
+			return fmt.Errorf("Compact API call failed: %v", err)
+		}
+		return nil
+	}(); err != nil {
+		return err
+	}
+	r.logger.Infof("Successfully compacted embedded etcd till revision: %v", revision)
+
+	ctx, cancel := context.WithTimeout(context.Background(), etcdConnectionTimeout)
```

Comment:
Multiple contexts are created in the `MakeEtcdLeanAndCheckAlarm` function. Could we simplify this by using a single context with a deadline for the entire process?

Reply:
parent context was there as I need to create context.
And multiple context are necessary according to the API calls like calling compaction might take time, hence context with bigger timeout.
Calling status API call shouldn't take more time, hence context with less timeout.